token/src/contract.rs

```diff
@@ -76,6 +41,51 @@ impl TokenTrait for Token {
         )
     }
 
+    pub fn clawback(e: Env, from: Address, amount: i128) {
+        check_nonnegative_amount(amount);
+        let admin = read_administrator(&e);
+        admin.require_auth();
+
+        e.storage().instance().bump(INSTANCE_BUMP_AMOUNT);
+
+        spend_balance(&e, from.clone(), amount);
+        event::clawback(&e, admin, from, amount);
+    }
+
+    pub fn set_authorized(e: Env, id: Address, authorize: bool) {
+        let admin = read_administrator(&e);
+        admin.require_auth();
+
+        e.storage().instance().bump(INSTANCE_BUMP_AMOUNT);
+
+        write_authorization(&e, id.clone(), authorize);
+        event::set_authorized(&e, admin, id, authorize);
+    }
+
+    pub fn mint(e: Env, to: Address, amount: i128) {
+        check_nonnegative_amount(amount);
+        let admin = read_administrator(&e);
+        admin.require_auth();
+
+        e.storage().instance().bump(INSTANCE_BUMP_AMOUNT);
+
+        receive_balance(&e, to.clone(), amount);
+        event::mint(&e, admin, to, amount);
+    }
+
+    pub fn set_admin(e: Env, new_admin: Address) {
+        let admin = read_administrator(&e);
+        admin.require_auth();
+
+        e.storage().instance().bump(INSTANCE_BUMP_AMOUNT);
+
+        write_administrator(&e, &new_admin);
+        event::set_admin(&e, admin, new_admin);
+    }
+}
+
+#[contractimpl]
+impl token::Interface for Token {
```

Comment:
It's unfortunate that we named the interface type simply `Interface`. Technically doing so adheres to some Rust best practice not to repeat the module name in the type. However, in importers we end up with simple `Interface` which does a poor job of communicating intent. Especially if there are multiple interfaces being implemented. It would have been better to name it `TokenInterface` I think.

So for the sake of readability importing it with the `token::` qualifier.

```rust
impl token::Interface for Token {
```

Note that this means that the import line is a little bonkers. It imports `token::self` so that `token` can be referenced. And it has to separately import `Interface` because implemented traits has to be imported into the scope to be used, otherwise the compiler will error on their use. The `as _` is so that `Interface` is imported into the scope without being given a name that must be used, since it is never referenced and would considered an unused import.

```rust
use soroban_sdk::token::{self, Interface as _};
```